# Initial data exploration for my datasets

More information about this project is available in my github repo here: https://github.com/Noah-Baustin/sf_crime_data_analysis

In [1]:
#import modules
import pandas as pd
#import altair as alt

/Users/nbaustin/.pyenv/versions/3.8.5/envs/sf_crime_data_analysis-3.8.5/lib/python3.8/site-packages/pandas/compat/__init__.py:124: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.
  warnings.warn(msg)


In [2]:
# import csv into a variable
historical_data = pd.read_csv('raw_data/SFPD_Incident_Reports_2003-May2018/Police_Department_Incident_Reports__Historical_2003_to_May_2018.csv', dtype=str)

Let's take a look at our data:

In [3]:
historical_data

PdId IncidntNum Incident Code        Category  \
0         3114751606302  031147516         06302   LARCENY/THEFT   
1         5069701104134  050697011         04134         ASSAULT   
2         6074729204104  060747292         04104         ASSAULT   
3         7103536315201  071035363         15201         ASSAULT   
4        11082415274000  110824152         74000  MISSING PERSON   
...                 ...        ...           ...             ...   
2129520  16099543005043  160995430         05043        BURGLARY   
2129521  16093783009031  160937830         09031  OTHER OFFENSES   
2129522  17078486916110  170784869         16110   DRUG/NARCOTIC   
2129523  18014823304138  180148233         04138         ASSAULT   
2129524  16044156851040  160441568         51040    NON-CRIMINAL   

                                                 Descript  DayOfWeek  \
0                             PETTY THEFT FROM A BUILDING     Sunday   
1                                                 BATTERY  Wednesday   
2                                                 ASSAULT   Saturday   
3                                                STALKING    Tuesday   
4                                           MISSING ADULT   Saturday   
...                                                   ...        ...   
2129520             BURGLARY OF RESIDENCE, UNLAWFUL ENTRY   Thursday   
2129521  MONEY, PROPERTY OR LABOR, FRAUDULENTLY OBTAINING     Monday   
2129522                    POSSESSION OF HEROIN FOR SALES    Tuesday   
2129523     BATTERY, FORMER SPOUSE OR DATING RELATIONSHIP   Saturday   
2129524                                        AIDED CASE     Monday   

               Date   Time PdDistrict      Resolution  ...  \
0        09/28/2003  10:00   SOUTHERN            NONE  ...   
1        06/22/2005  12:20   NORTHERN            NONE  ...   
2        07/15/2006  00:55    CENTRAL            NONE  ...   
3        09/25/2007  00:01    TARAVAL            NONE  ...   
4        09/24/2011  11:00    TARAVAL         LOCATED  ...   
...             ...    ...        ...             ...  ...   
2129520  12/01/2016  17:30   NORTHERN            NONE  ...   
2129521  11/14/2016  00:01  INGLESIDE            NONE  ...   
2129522  09/26/2017  01:38    CENTRAL  ARREST, BOOKED  ...   
2129523  02/24/2018  20:30    CENTRAL  ARREST, BOOKED  ...   
2129524  05/30/2016  11:33   SOUTHERN            NONE  ...   

        Fix It Zones as of 2017-11-06  2 2 DELETE - HSOC Zones 2 2  \
0                                      NaN                     NaN   
1                                      NaN                     NaN   
2                                      NaN                     NaN   
3                                      NaN                     NaN   
4                                      NaN                     NaN   
...                                    ...                     ...   
2129520                                NaN                       1   
2129521                                NaN                     NaN   
2129522                                 23                     NaN   
2129523                                NaN                     NaN   
2129524                                NaN                       1   

        Fix It Zones as of 2018-02-07 2 2  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   
...                                   ...   
2129520                               NaN   
2129521                               NaN   
2129522                                23   
2129523                               NaN   
2129524                               NaN   

        CBD, BID and GBD Boundaries as of 2017 2 2  \
0                                              NaN   
1                                              NaN   
2                                              N

In [4]:
historical_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2129525 entries, 0 to 2129524
Data columns (total 35 columns):
 #   Column                                                    Dtype 
---  ------                                                    ----- 
 0   PdId                                                      object
 1   IncidntNum                                                object
 2   Incident Code                                             object
 3   Category                                                  object
 4   Descript                                                  object
 5   DayOfWeek                                                 object
 6   Date                                                      object
 7   Time                                                      object
 8   PdDistrict                                                object
 9   Resolution                                                object
 10  Address                                   

I also see that I've got some date columns that need to be reformatted: for sure 'Date'. Let's reformat that:

In [5]:
historical_data['Date'] = pd.to_datetime(historical_data['Date'])

In [6]:
historical_data.head()

PdId IncidntNum Incident Code        Category  \
0   3114751606302  031147516         06302   LARCENY/THEFT   
1   5069701104134  050697011         04134         ASSAULT   
2   6074729204104  060747292         04104         ASSAULT   
3   7103536315201  071035363         15201         ASSAULT   
4  11082415274000  110824152         74000  MISSING PERSON   

                      Descript  DayOfWeek       Date   Time PdDistrict  \
0  PETTY THEFT FROM A BUILDING     Sunday 2003-09-28  10:00   SOUTHERN   
1                      BATTERY  Wednesday 2005-06-22  12:20   NORTHERN   
2                      ASSAULT   Saturday 2006-07-15  00:55    CENTRAL   
3                     STALKING    Tuesday 2007-09-25  00:01    TARAVAL   
4                MISSING ADULT   Saturday 2011-09-24  11:00    TARAVAL   

  Resolution  ... Fix It Zones as of 2017-11-06  2 2 DELETE - HSOC Zones 2 2  \
0       NONE  ...                                NaN                     NaN   
1       NONE  ...                                NaN                     NaN   
2       NONE  ...                                NaN                     NaN   
3       NONE  ...                                NaN                     NaN   
4    LOCATED  ...                                NaN                     NaN   

  Fix It Zones as of 2018-02-07 2 2  \
0                               NaN   
1                               NaN   
2                               NaN   
3                               NaN   
4                               NaN   

  CBD, BID and GBD Boundaries as of 2017 2 2 Areas of Vulnerability, 2016 2 2  \
0                                        NaN                              NaN   
1                                        NaN                                2   
2                                        NaN                                2   
3                                        NaN                                1   
4                                        NaN                              NaN   

  Central Market/Tenderloin Boundary 2 2  \
0                                    NaN   
1                                    NaN   
2                                    NaN   
3                                    NaN   
4                                    NaN   

  Central Market/Tenderloin Boundary Polygon - Updated 2 2  \
0                                                NaN         
1                                                NaN         
2                                                NaN         
3                                                NaN         
4                                                NaN         

  HSOC Zones as of 2018-06-05 2 2 OWED Public Spaces 2 2 Neighborhoods 2  
0                             NaN                    NaN             NaN  
1                             NaN                    NaN              97  
2                             NaN                    NaN             106  
3                             NaN                    NaN              49  
4                             NaN                    NaN             NaN  

[5 rows x 35 columns]

Now that our dates are formatted correctly, let's take a look at our date range included in the data:

In [7]:
historical_data['Date'].min()

Timestamp('2003-01-01 00:00:00')

In [8]:
historical_data['Date'].max()

Timestamp('2018-05-15 00:00:00')

We can see here that it appears that we most likely have complete data beginning in 2003. BUT we see here that the 2018 data is incomplete, so if we want to do an annual analysis, we'll need to exclude 2018.

We've got 2,129,525 entries. Let's check out our three different columns that have incident indentification codes to see how many unique values are in each:

In [9]:
historical_data['IncidntNum'].nunique()

1703626

In [15]:
historical_data['Incident Code'].nunique()

885

In [16]:
historical_data['PdId'].nunique()

2129525

Our documentation tells us that the PdId column is equivelant to the row_id column in the new data (see downloaded pdf titled: Change Notice - Police Incident Reports). And the documentation for the newer dataset tells us that the row_id is the unique identifier for each row. 

It's a good sign that there's exaclty as many PdId unique values as there are rows in the dataset.

It also makes sense that there are duplicate values in the IncidntNum column. The IncidntNum refers to the case number. So if a supplemental report was filed after the incident was initially entered in this dataset, it would show up with a new PdId BUT the IncidntNum would be the same. 

But that does mean that I need to make sure I'm not counting the same incident multiple times if it shows up in this dataset multiple times. Most likely, my analysis will focus on a unique set of IncidntNum. 

Let's double check that there's no duplicate rows:

In [19]:
historical_data[historical_data.duplicated()]

Empty DataFrame
Columns: [PdId, IncidntNum, Incident Code, Category, Descript, DayOfWeek, Date, Time, PdDistrict, Resolution, Address, X, Y, location, SF Find Neighborhoods 2 2, Current Police Districts 2 2, Current Supervisor Districts 2 2, Analysis Neighborhoods 2 2, DELETE - Fire Prevention Districts 2 2, DELETE - Police Districts 2 2, DELETE - Supervisor Districts 2 2, DELETE - Zip Codes 2 2, DELETE - Neighborhoods 2 2, DELETE - 2017 Fix It Zones 2 2, Civic Center Harm Reduction Project Boundary 2 2, Fix It Zones as of 2017-11-06  2 2, DELETE - HSOC Zones 2 2, Fix It Zones as of 2018-02-07 2 2, CBD, BID and GBD Boundaries as of 2017 2 2, Areas of Vulnerability, 2016 2 2, Central Market/Tenderloin Boundary 2 2, Central Market/Tenderloin Boundary Polygon - Updated 2 2, HSOC Zones as of 2018-06-05 2 2, OWED Public Spaces 2 2, Neighborhoods 2]
Index: []

[0 rows x 35 columns]

No duplicate rows, that's great.

Let's find out how many duplicated IncidntNum values we have:

In [20]:
historical_data[historical_data['IncidntNum'].duplicated()]

PdId IncidntNum Incident Code        Category  \
523      17001676971010  170016769         71010  OTHER OFFENSES   
562      18023965028150  180239650         28150       VANDALISM   
857      17018698663010  170186986         63010        WARRANTS   
968      17063457109320  170634571         09320           FRAUD   
1179     16096494626135  160964946         26135         ASSAULT   
...                 ...        ...           ...             ...   
2129517  16092245574000  160922455         74000  MISSING PERSON   
2129519  18014350616652  180143506         16652   DRUG/NARCOTIC   
2129521  16093783009031  160937830         09031  OTHER OFFENSES   
2129522  17078486916110  170784869         16110   DRUG/NARCOTIC   
2129524  16044156851040  160441568         51040    NON-CRIMINAL   

                                                 Descript DayOfWeek  \
523                             LOST/STOLEN LICENSE PLATE  Thursday   
562                         MALICIOUS MISCHIEF, VANDALISM  Saturday   
857                                        WARRANT ARREST    Monday   
968                          CREDIT CARD, THEFT BY USE OF    Friday   
1179         UNLAWFUL DISSUADING/THREATENING OF A WITNESS    Sunday   
...                                                   ...       ...   
2129517                                     MISSING ADULT  Saturday   
2129519           POSSESSION OF METH-AMPHETAMINE FOR SALE  Thursday   
2129521  MONEY, PROPERTY OR LABOR, FRAUDULENTLY OBTAINING    Monday   
2129522                    POSSESSION OF HEROIN FOR SALES   Tuesday   
2129524                                        AIDED CASE    Monday   

              Date   Time  PdDistrict      Resolution  ...  \
523     2017-01-05  18:00     BAYVIEW            NONE  ...   
562     2018-03-31  14:30     TARAVAL            NONE  ...   
857     2017-03-06  16:00  TENDERLOIN  ARREST, BOOKED  ...   
968     2017-08-04  16:00     CENTRAL            NONE  ...   
1179    2016-11-27  14:29     BAYVIEW  ARREST, BOOKED  ...   
...            ...    ...         ...             ...  ...   
2129517 2016-11-12  15:08    SOUTHERN            NONE  ...   
2129519 2018-02-22  23:27     CENTRAL  ARREST, BOOKED  ...   
2129521 2016-11-14  00:01   INGLESIDE            NONE  ...   
2129522 2017-09-26  01:38     CENTRAL  ARREST, BOOKED  ...   
2129524 2016-05-30  11:33    SOUTHERN            NONE  ...   

        Fix It Zones as of 2017-11-06  2 2 DELETE - HSOC Zones 2 2  \
523                                    NaN                     NaN   
562                                    NaN                     NaN   
857                                     18                     NaN   
968                                     12                     NaN   
1179                                   NaN                     NaN   
...                                    ...                     ...   
2129517                                NaN                     NaN   
2129519                                 12                     NaN   
2129521                                NaN                     NaN   
2129522                                 23                     NaN   
2129524                                NaN                       1   

        Fix It Zones as of 2018-02-07 2 2  \
523                                   NaN   
562                                   NaN   
857                                    18   
968                                    12   
1179                                  NaN   
...                                   ...   
2129517                               NaN   
2129519                                12   
2129521                               NaN   
2129522                                23   
2129524                               NaN   

        CBD, BID and GBD Boundaries as of 2017 2 2  \
523                                            NaN   
562                                            NaN   
857                                              6   
968   

In [21]:
1703626 + 425899

2129525

There's 425,899 duplicates for 'IncidntNum'. We we add that to the number of unique values (above) it is equal to the number of rows in our dataset... that tells us that 1,703,626 is the actual number of incidents that we're working with.

Let's take a look at some of our duplicates:

First we'll create a dataframe with our duplicated cases:

In [29]:
dupe_cases = historical_data[historical_data['IncidntNum'].duplicated()].copy()

Now we create a `list` or `array` (the numpy version of a list) of those unique PdId's:

In [40]:
dupe_IncidentNum = dupe_cases['IncidntNum'].to_list()

Finally, we're displaying a subset our of original data that just includes the duplicate incident numbers:

In [43]:
dupe_cases_full = historical_data[historical_data['IncidntNum'].isin(dupe_IncidentNum)].sort_values(by='IncidntNum')

In [44]:
dupe_cases_full.head(30)

PdId IncidntNum Incident Code                Category  \
1021616    6238907023  000062389         07023           VEHICLE THEFT   
1869704    6238907043  000062389         07043           VEHICLE THEFT   
1003273   12343616010  000123436         16010           DRUG/NARCOTIC   
625462    12343602004  000123436         02004  SEX OFFENSES, FORCIBLE   
1836290   24825063010  000248250         63010                WARRANTS   
1009328   24825062050  000248250         62050                WARRANTS   
1554935   32468419090  000324684         19090             DRUNKENNESS   
1297782   32468463010  000324684         63010                WARRANTS   
1434901   36322862050  000363228         62050                WARRANTS   
1029718   36322863010  000363228         63010                WARRANTS   
797056    76724426080  000767244         26080          OTHER OFFENSES   
1734920   76724416621  000767244         16621           DRUG/NARCOTIC   
1238053   76724416700  000767244         16700           DRUG/NARCOTIC   
1279759   81020726142  000810207         26142              KIDNAPPING   
444741    81020704144  000810207         04144  SEX OFFENSES, FORCIBLE   
1362027   81020702204  000810207         02204  SEX OFFENSES, FORCIBLE   
490689    88865207023  000888652         07023           VEHICLE THEFT   
464553    88865207043  000888652         07043           VEHICLE THEFT   
802280    96275107023  000962751         07023           VEHICLE THEFT   
870203    96275107043  000962751         07043           VEHICLE THEFT   
1373007  100017663010  001000176         63010                WARRANTS   
1227820  100017662050  001000176         62050                WARRANTS   
484782   124909527065  001249095         27065                   FRAUD   
886947   124909563010  001249095         63010                WARRANTS   
1793335  132654662020  001326546         62020                WARRANTS   
1563382  132654663010  001326546         63010                WARRANTS   
881525   144089562060  001440895         62060                WARRANTS   
1711046  144089563010  001440895         63010                WARRANTS   
1892575  313990175000  003139901         75000          MISSING PERSON   
893672   313990174000  003139901         74000          MISSING PERSON   

                                             Descript  DayOfWeek       Date  \
1021616                             STOLEN MOTORCYCLE    Tuesday 2004-03-30   
1869704                VEHICLE, RECOVERED, MOTORCYCLE    Tuesday 2004-03-30   
1003273                       POSSESSION OF MARIJUANA     Friday 2004-07-02   
625462                    FORCIBLE RAPE, BODILY FORCE     Friday 2004-07-02   
1836290                                WARRANT ARREST     Monday 2003-01-27   
1009328               ENROUTE TO OUTSIDE JURISDICTION     Monday 2003-01-27   
1554935  UNDER INFLUENCE OF ALCOHOL IN A PUBLIC PLACE     Monday 2003-03-31   
1297782                                WARRANT ARREST     Monday 2003-03-31   
1434901               ENROUTE TO OUTSIDE JURISDICTION  Wednesday 2004-11-03   
1029718                                WARRANT ARREST  Wednesday 2004-11-03   
797056                                     CONSPIRACY     Sunday 2003-06-29   
1734920               POSSESSION OF BASE/ROCK COCAINE     Sunday 2003-06-29   
1238053       VISITING WHERE DRUGS ARE USED OR SMOKED     Sunday 2003-06-29   
1279759                            FALSE IMPRISONMENT  Wednesday 2003-07-09   
444741                                 SEXUAL BATTERY  Wednesday 2003-07-09   
1362027                  ATTEMPTED RAPE, BODILY FORCE  Wednesday 2003-07-09   
490689                              STOLEN MOTORCYCLE    Tuesday 2004-06-15   
464553                 VEHICLE, RECOVERED, MOTORCYCLE    Tuesday 2004-06-15   
802280                              STOLEN MOTORCYCLE  Wednesday 2004-06-23   
870203                 VEHICLE, RECOVERED, MOTORCYCLE  Wednesday 2004-06-23   
1373007                                WARRANT ARREST     Monda

We can see here anecdotely that additional entries for duplicated IncidentNum's include additional entries in the descript column in some cases, but sometimes they remain the same. 
#### come back to this!

We're going to want to isolate our incidents that include marijuana crimes. So let's take a look at the unique values in the two columns that might contain information about marijuana crimes:

In [45]:
historical_data['Category'].unique()

array(['LARCENY/THEFT', 'ASSAULT', 'MISSING PERSON', 'VEHICLE THEFT',
       'BURGLARY', 'DRUG/NARCOTIC', 'DRIVING UNDER THE INFLUENCE',
       'VANDALISM', 'OTHER OFFENSES', 'DRUNKENNESS', 'NON-CRIMINAL',
       'ROBBERY', 'SUSPICIOUS OCC', 'TRESPASS', 'WARRANTS',
       'FORGERY/COUNTERFEITING', 'STOLEN PROPERTY',
       'SEX OFFENSES, FORCIBLE', 'FRAUD', 'SECONDARY CODES',
       'PROSTITUTION', 'RECOVERED VEHICLE', 'BRIBERY', 'ARSON',
       'DISORDERLY CONDUCT', 'WEAPON LAWS', 'LIQUOR LAWS', 'EXTORTION',
       'SUICIDE', 'KIDNAPPING', 'SEX OFFENSES, NON FORCIBLE',
       'BAD CHECKS', 'EMBEZZLEMENT', 'LOITERING', 'GAMBLING', 'TREA',
       'PORNOGRAPHY/OBSCENE MAT'], dtype=object)

We can see here that there's a DRUG/NARCOTIC category, so that's probably where we're going to find the marijuana crimes. But nothing specific about marijuana here. That's going to show up in our 'Descript' column. There's too many unique values in that column to list, so let's create a subset:

In [25]:
historical_data_marijuana = historical_data[
    historical_data['Descript'].str.contains('MARIJUANA')
].copy()

 I also see that there's a bunch of columns in the data that are not included in the data dictionary on the city site where we got the data from: https://data.sfgov.org/Public-Safety/Police-Department-Incident-Reports-Historical-2003/tmnf-yvry
 
 #### Eventually: will want to get rid of these spare columns!!